src/librustc/ty/sty.rs

```diff
@@ -571,14 +571,28 @@ impl<'a, 'tcx> ProjectionTy<'tcx> {
             |item| item.name == item_name).unwrap().def_id;
 
         ProjectionTy {
-            trait_ref: trait_ref,
+            substs: trait_ref.substs,
             item_def_id: item_def_id,
         }
     }
 
-    pub fn item_name(self, tcx: TyCtxt) -> Name {
+    pub fn item_name(&self, tcx: TyCtxt) -> Name {
         tcx.associated_item(self.item_def_id).name
     }
+
+    pub fn trait_ref(&self, tcx: TyCtxt) -> ty::TraitRef<'tcx> {
+        let def_id = match tcx.associated_item(self.item_def_id).container {
+                ty::TraitContainer(def_id) => def_id,
+                // FIXME(tschottdorf): this case is impossible.
+                ty::ImplContainer(def_id) => {
+                    bug!("asked for a TraitContainer, got an ImplContainer with DefId {:?}", def_id)
+                }
```

Comment:
Irrelevant and there is a method to get the `DefId` anyway, that'd make this cleaner.